### PR TITLE
Avoid rpc_mutex deadlock

### PR DESCRIPTION
**Pull request by linuxsmiths**
rpc_set_error() holds the rpc_mutex so it cannot be called by any function which already holds the mutex. In this case rpc_write_to_socket() was calling rpc_set_error() resulting in a deadlock. This was detected when running the stress test suite where I continously keep killing the connectiong with "ss -K" while data xfer is on.

To solve this I defined a nolock version for both the rpc and nfs set error functions. Audited the code and fixed one more occurrence. The audit may not be complete so created "error checking" mutex to catch such locking violations.

Didn't use recursive mutex as those encourage bad programming practice and such bugs may silently get hidden.

Testing done:
With this fix ran 100+ iterations of 100GB file xfer while killing all connections every 5 seconds.

### CODE REVIEW

**Comment by linuxsmiths** [comment]
@sahlberg, I've done basic audit, so if you get a chance pls audit to see if we are calling rpc_set_error()/nfs_set_error() in contexts already holding the rpc_mutex.

fyi, I'll be away travelling for next couple of days so my responses may be delayed.

File: lib/init.c [hunk 1]
```diff
@@ -296,31 +296,41 @@ void rpc_set_auxiliary_gids(struct rpc_context *rpc, uint32_t len, uint32_t* gid
 void rpc_set_error(struct rpc_context *rpc, const char *error_string, ...)
 {
         va_list ap;
-	char *old_error_string;
```
Comment (sahlberg) [inline lib/init.c:299]: I don't think we can do this change and drop old_error_string.
The reason is it is useful to be able to append the previous error strong to the new one being generated.

For example instances like this:
                smb2_set_error(smb2, "Failed to create set command. %s",
                               smb2_get_error(smb2));

So undo this change.

File: lib/init.c [hunk 2]
```diff
@@ -296,31 +296,41 @@ void rpc_set_auxiliary_gids(struct rpc_context *rpc, uint32_t len, uint32_t* gid
...
-		free(old_error_string);
-	}
 #ifdef HAVE_MULTITHREADING
         if (rpc->multithreading_enabled) {
                 nfs_mt_mutex_unlock(&rpc->rpc_mutex);
         }
 #endif /* HAVE_MULTITHREADING */
 }
 
+void rpc_set_error_nolock(struct rpc_context *rpc, const char *error_string, ...)
```
Comment (sahlberg) [inline lib/init.c:321]: See the previous comment on why we need the old_error_string variable here and why we need to defer freeing it until after we have created the new string in vsnprintf.

File: lib/init.c [hunk 3]
```diff
@@ -296,31 +296,41 @@ void rpc_set_auxiliary_gids(struct rpc_context *rpc, uint32_t len, uint32_t* gid
...
 
 #ifdef HAVE_MULTITHREADING
         if (rpc->multithreading_enabled) {
                 nfs_mt_mutex_lock(&rpc->rpc_mutex);
         }
 #endif /* HAVE_MULTITHREADING */
-        old_error_string = rpc->error_string;
+
         va_start(ap, error_string);
+	free(rpc->error_string);
```
Comment (sahlberg) [inline lib/init.c:307]: Ideally this whole function should just be the locking and then a call to the _locked() function
so that we only have one copy of the code that generates the new error string.

File: lib/init.c (hunk 2 above)
Comment (sahlberg) [inline lib/init.c:321]: Change the name to rpc_set_error_locked() to indicate this is called when the lock is already held.

**Review by sahlberg — CHANGES_REQUESTED** [review]

File: lib/init.c (hunk 1 above)
Comment (linuxsmiths) [reply lib/init.c:299]: good point! will update

File: lib/init.c (hunk 3 above)
Comment (linuxsmiths) [reply lib/init.c:307]: yes, I thought too but due to the varargs we have to either take a va_list as argument in which case the locked() version cannot be directly called or we make that a macro. For now, I'll leave it as is.
